packages/ui/src/components/form-primitives/select.tsx

```diff
@@ -0,0 +1,425 @@
+  ReactElement,
+  ReactNode,
+  UIEvent,
+  useCallback,
+  useEffect,
+  useMemo,
+  useRef,
+  useState
+} from 'react'
+
+import { ControlGroup, FormCaption, IconV2, Label, SearchInput } from '@/components'
+import { useTranslation } from '@/context'
+import { cn, generateAlphaNumericHash } from '@/utils'
+import { DropdownMenu } from '@components/dropdown-menu'
+import { Text } from '@components/text'
+import { cva, VariantProps } from 'class-variance-authority'
+import debounce from 'lodash-es/debounce'
+
+const selectVariants = cva('cn-select', {
+  variants: {
+    theme: {
+      default: '',
+      danger: 'cn-select-danger',
+      warning: 'cn-select-warning'
+    }
+  },
+  defaultVariants: {
+    theme: 'default'
+  }
+})
+
+type SeparatorOption = '-'
+
+interface ValueOption<T = string> {
+  label: string | ReactNode
+  value: T
+  disabled?: boolean
+}
+
+interface GroupOption<T = string> {
+  label: string
+  options: Array<ValueOption<T> | SeparatorOption>
+}
+
+type SelectOption<T = string> = ValueOption<T> | GroupOption<T> | SeparatorOption
+
+type SelectItemType =
+  | ReactElement<typeof DropdownMenu.Item>
+  | ReactElement<typeof DropdownMenu.AvatarItem>
+  | ReactElement<typeof DropdownMenu.IconItem>
+  | ReactElement<typeof DropdownMenu.LogoItem>
+  | ReactElement<typeof DropdownMenu.IndicatorItem>
+
+interface SelectProps<T = string>
+  extends Pick<InputHTMLAttributes<HTMLInputElement>, 'id' | 'name' | 'disabled' | 'placeholder' | 'className'> {
+  options: SelectOption<T>[] | (() => Promise<SelectOption<T>[]>)
+  value?: T
+  defaultValue?: T
+  onChange?: (value: T) => void
+  onScrollEnd?: () => void
+  isLoading?: boolean
+  label?: string
+  theme?: VariantProps<typeof selectVariants>['theme']
+  caption?: string
+  error?: string
+  warning?: string
+  optional?: boolean
+  allowSearch?: boolean
+  onSearch?: (query: string) => void
+  searchValue?: string
+  optionRenderer?: (option: ValueOption<T>) => SelectItemType
+}
+
+// Helper function to check option types
+const isValueOption = <T,>(option: SelectOption<T>): option is ValueOption<T> => {
+  return typeof option === 'object' && 'value' in option
+}
+
+const isGroupOption = <T,>(option: SelectOption<T>): option is GroupOption<T> => {
+  return typeof option === 'object' && 'options' in option
+}
+
+const isSeparatorOption = <T,>(option: SelectOption<T>): option is SeparatorOption => {
+  return option === '-'
+}
+
+// Helper to get all value options (flattened)
+const getAllValueOptions = <T,>(options: SelectOption<T>[]): ValueOption<T>[] => {
+  const valueOptions: ValueOption<T>[] = []
+
+  options.forEach(option => {
+    if (isValueOption(option)) {
+      valueOptions.push(option)
+    } else if (isGroupOption(option)) {
+      option.options.forEach(subOption => {
+        if (isValueOption(subOption)) {
+          valueOptions.push(subOption)
+        }
+      })
+    }
+  })
+
+  return valueOptions
+}
+
+function SelectInner<T = string>(
```

Comment:
Please add a backlog ticket to add `forwardRef` for `Select` component.

Reply:
forwardRef is added on line 414

```js
const Select = forwardRef(SelectInner) as <T = string>(
  props: SelectProps<T> & { ref?: ForwardedRef<HTMLButtonElement> }
) => ReturnType<typeof SelectInner>
```